vendor/github.com/openshift/library-go/pkg/monitor/health/prober.go

```diff
@@ -0,0 +1,377 @@
+}
+
+// refreshTargetsLocked updates the internal targets list to monitor if it was requested (via the Enqueue method)
+func (sm *Prober) refreshTargetsLocked() {
+	sm.refreshTargetsLock.Lock()
+	defer sm.refreshTargetsLock.Unlock()
+	if !sm.refreshTargets {
+		return
+	}
+
+	sm.refreshTargets = false
+	freshTargets := sm.targetProvider.CurrentTargetsList()
+	freshTargetSet := sets.NewString(freshTargets...)
+
+	currentTargetsSet := sets.NewString(sm.targetsToMonitor...)
+	newTargetsToMonitorSet := freshTargetSet.Difference(currentTargetsSet)
+	if newTargetsToMonitorSet.Len() > 0 {
+		klog.V(2).Infof("health monitor observed new targets = %v", newTargetsToMonitorSet.List())
+	}
+
+	removedTargetsToMonitorSet := currentTargetsSet.Difference(freshTargetSet)
+	if removedTargetsToMonitorSet.Len() > 0 {
+		klog.V(2).Infof("health monitor will stop checking the following targets targets = %v", removedTargetsToMonitorSet.List())
+		for targetToRemove := range removedTargetsToMonitorSet {
+			delete(sm.consecutiveSuccessfulProbes, targetToRemove)
+			delete(sm.consecutiveFailedProbes, targetToRemove)
+		}
+
+		healthyTargetsSet := sets.NewString(sm.healthyTargets...)
+		healthyTargetsSet.Delete(removedTargetsToMonitorSet.List()...)
+		sm.healthyTargets = healthyTargetsSet.List()
+
+		unhealthyTargetsSet := sets.NewString(sm.unhealthyTargets...)
+		unhealthyTargetsSet.Delete(removedTargetsToMonitorSet.List()...)
+		sm.unhealthyTargets = unhealthyTargetsSet.List()
+	}
+
+	sm.targetsToMonitor = freshTargets
+}
+
+func (sm *Prober) healthCheckRegisteredTargets() {
+	sm.refreshTargetsLocked()
+	var wg sync.WaitGroup
+	resTargetErrTupleCh := make(chan targetErrTuple, len(sm.targetsToMonitor))
+
+	for i := 0; i < len(sm.targetsToMonitor); i++ {
+		wg.Add(1)
+		go func(target string) {
+			defer wg.Done()
+			err := sm.healthCheckSingleTarget(target)
+			resTargetErrTupleCh <- targetErrTuple{target, err}
+		}(sm.targetsToMonitor[i])
+	}
+	wg.Wait()
+	close(resTargetErrTupleCh)
+
+	currentHealthCheckProbes := make([]targetErrTuple, 0, len(sm.targetsToMonitor))
+	for svrErrTuple := range resTargetErrTupleCh {
+		currentHealthCheckProbes = append(currentHealthCheckProbes, svrErrTuple)
+	}
+
+	sm.updateHealthChecksFor(currentHealthCheckProbes)
+}
+
+// updateHealthChecksFor examines the health of targets based on the provided probes and the current configuration.
+// It also notifies interested parties about changes in the health condition.
+// Interested parties can be registered by calling AddListener method.
+func (sm *Prober) updateHealthChecksFor(currentHealthCheckProbes []targetErrTuple) {
+	newUnhealthyTargets := []string{}
+	newHealthyTargets := []string{}
+
+	for _, svrErrTuple := range currentHealthCheckProbes {
+		if svrErrTuple.err != nil {
+			delete(sm.consecutiveSuccessfulProbes, svrErrTuple.target)
+
+			unhealthyProbesSlice := sm.consecutiveFailedProbes[svrErrTuple.target]
+			if len(unhealthyProbesSlice) < sm.unhealthyProbesThreshold {
```

Comment:
`if len(unhealthyProbesSlice) > sm.unhealthyProbesThreshold `  don't we want to keep the most recent errors?
this way we only have the errors that make the probe fail, but we don't know what is causing the probes failures at this point in time

Reply:
we do log unhealthy targets in `logUnhealthyTargets`